// wm/renderer/ui/Ui.js
var ReactUi = require('wm/renderer/ui/ReactUi');

function Ui(config){
    Object.assign(this, config);
    if(!this.logicBus) throw new Error('No logicBus object specified for Ui!');
    if(!this.core) throw new Error('No core object specified for Ui!');

    //var reactUi = new ReactUi(); 
    this.makeUi();
}

Ui.prototype.makeUi = function(){
    var startButton = document.getElementById('startButton');
    startButton.addEventListener('click', this.onStartButtonClick.bind(this, startButton));
};

Ui.prototype.onStartButtonClick = function(){
    this.logicBus.postMessage('start',{});
    this.core.startGameRenderMode();

    var viewportElement = document.getElementById('viewport');

    viewportElement.removeClass('blur');
    viewportElement.addClass('blurEnd');

    viewportElement.addEventListener('animationend', function(){
        viewportElement.removeClass('blurEnd');
    });

    this.hide(document.getElementById('startScreen'));
};

Ui.prototype.hide = function(element){
    element.style.display = 'none';
};

Ui.prototype.show = function(element){
    element.style.display = '';
};

Ui.prototype.showStopGame = function(info){
    document.getElementById('viewport').addClass('blurStart');
    document.getElementById('endText').addClass('textAppear');
    document.getElementById('scoreText').addClass('textAppear');
    document.getElementById('scoreText').innerHTML = 'KILLED: ' + info.killed + '<br>REMAINING: ' + info.remaining + '<br><br>' + this.getOpinionOnResult(info.remaining);
    this.show(document.getElementById('endScreen'));
};

Ui.prototype.getOpinionOnResult = function(remainingMooks){
    if (remainingMooks === 100){
        return 'You didn\'t even try, did you?';
    } else if (remainingMooks > 90 && remainingMooks < 100){
        return 'You seem to have discovered shooting function.';
    } else if (remainingMooks > 80 && remainingMooks < 90){
        return 'Far, far away.';
    } else if (remainingMooks > 70 && remainingMooks < 80){
        return 'Come on. You can do better! I hope, for this is only a techtest and they still suck.';
    } else if (remainingMooks > 60 && remainingMooks < 70){
        return 'Try using your second weapon on them. Works much better.';
    } else if (remainingMooks > 50 && remainingMooks < 60){
        return 'You know you can shoot down these orange blobs with your primary weapon?';
    } else if (remainingMooks > 40 && remainingMooks < 50){
        return 'Only half more to go.';
    } else if (remainingMooks > 30 && remainingMooks < 40){
        return 'That is a formidable effort.';
    } else if (remainingMooks > 20 && remainingMooks < 30){
        return 'It should be getting easier by now.';
    } else if (remainingMooks > 10 && remainingMooks < 20){
        return 'So close.';
    } else if (remainingMooks > 0 && remainingMooks < 10){
        return 'Almost there. Got unlucky with a stray shot?';
    } else {
        return 'You got them all! Grats!';
    }
};


module.exports = Ui;
