// wm/renderer/actorManagement/ActorFactory.js
var idMap = {
    SHIP: 1,
    MOOK: 2,
    PILLAR: 3,
    WALL: 4,
    CHUNK: 5,
    PLASMAPROJECTILE: 100,
    MOLTENPROJECTILE: 101,
    LASERPROJECITLE: 102,
};

function ActorFactory(actorDependencies){
    this.actorDependencies = actorDependencies;
    this.actorMap = {
        [idMap.SHIP]: ActorFactory.ShipActor,
        [idMap.MOOK]: ActorFactory.MookActor,
        [idMap.WALL]: ActorFactory.WallActor,
        [idMap.PILLAR]: ActorFactory.PillarActor,
        [idMap.CHUNK]: ActorFactory.ChunkActor,
        [idMap.PLASMAPROJECTILE]: ActorFactory.PlasmaProjectileActor,
        [idMap.MOLTENPROJECTILE]: ActorFactory.MoltenProjectileActor,
        [idMap.LASERPROJECITLE]: ActorFactory.LaserProjectileActor
    };
}

ActorFactory.prototype.create = function(config){
    if(!this.actorMap[config.classId]){
        throw new Error("Cannot create actor. Bad configuration!". config);
    }
    return new this.actorMap[config.classId](config, this.actorDependencies);
};

module.exports = function(context){
    ActorFactory.ShipActor = context === 'renderer' ? require("wm/renderer/actor/player/ShipActor") : require("wm/logic/actor/player/ShipActor");
    ActorFactory.MookActor = context === 'renderer' ? require("wm/renderer/actor/enemy/MookActor") : require("wm/logic/actor/enemy/MookActor");
    ActorFactory.WallActor = context === 'renderer' ? require("wm/renderer/actor/map/WallActor") : require("wm/logic/actor/map/WallActor");
    ActorFactory.PillarActor = context === 'renderer' ? require("wm/renderer/actor/map/PillarActor") : require("wm/logic/actor/map/PillarActor");
    ActorFactory.ChunkActor = context === 'renderer' ? require("wm/renderer/actor/object/ChunkActor") : require("wm/logic/actor/object/ChunkActor");
    ActorFactory.PlasmaProjectileActor = context === 'renderer' ? require("wm/renderer/actor/projectile/PlasmaProjectileActor") : require("wm/logic/actor/projectile/PlasmaProjectileActor");
    ActorFactory.MoltenProjectileActor = context === 'renderer' ? require("wm/renderer/actor/projectile/MoltenProjectileActor") : require("wm/logic/actor/projectile/MoltenProjectileActor");
    ActorFactory.LaserProjectileActor = context === 'renderer' ? require("wm/renderer/actor/projectile/LaserProjectileActor") : require("wm/logic/actor/projectile/LaserProjectileActor");

    var returnObject = {};

    returnObject.getInstance = function(dependencies){
        return new ActorFactory(dependencies);
    };

    Object.keys(idMap).forEach(function(key){
        returnObject[key] = idMap[key];
    });

    return returnObject;
};
